src/commissioner/Events.h

```diff
@@ -0,0 +1,274 @@
+        if (mData.get() == nullptr)
+        {
+            return CHIP_ERROR_NO_MEMORY;
+        }
+        mLen      = 0;
+        mCapacity = size;
+        return CHIP_NO_ERROR;
+    }
+
+private:
+    Platform::SharedPtr<uint8_t> mData;
+    size_t mCapacity = 0;
+    size_t mLen      = 0;
+};
+
+template <size_t MaxSize>
+struct CapacityBoundSharedBuffer : public SharedBuffer
+{
+    CHIP_ERROR Allocate(size_t size)
+    {
+        if (size > MaxSize)
+        {
+            return CHIP_ERROR_NO_MEMORY;
+        }
+        return SharedBuffer::Allocate(size);
+    }
+    CHIP_ERROR Set(const ByteSpan buffer)
+    {
+        if (buffer.size() > MaxSize)
+        {
+            return CHIP_ERROR_NO_MEMORY;
+        }
+        return SharedBuffer::Set(buffer);
+    }
+};
+
+template <size_t MaxSize>
+struct CapacityBoundStaticBuffer
+{
+    const ByteSpan Get() const { return ByteSpan(mBuffer, mBuffer.Length()); }
+    void Clear() { mBuffer.SetLength(0); };
+    CHIP_ERROR Set(ByteSpan buffer)
+    {
+        ReturnErrorOnFailure(mBuffer.SetLength(buffer.size()));
+        memmove(mBuffer, buffer.data(), buffer.size());
+        return CHIP_NO_ERROR;
+    }
+
+private:
+    Crypto::CapacityBoundBuffer<MaxSize> mBuffer;
+};
+
+// Events
+struct Success
+{
+};
+
+struct Failure
+{
+};
+
+struct Await
+{
+};
+
+struct OnboardingPayload
+{
+    const char * mPayload;
+};
+
+struct NetworkId : CapacityBoundStaticBuffer<kMaxNetworkIdLength>
+{
+    CHIP_ERROR FromThreadOperationalDataset(ByteSpan opaqueDataset)
+    {
+        Thread::OperationalDataset parsed;
+        ReturnErrorOnFailure(parsed.Init(opaqueDataset));
+        uint8_t extendedPanId[Thread::kSizeExtendedPanId];
+        ReturnErrorOnFailure(parsed.GetExtendedPanId(extendedPanId));
+        ReturnErrorOnFailure(Set(ByteSpan(extendedPanId)));
+        return CHIP_NO_ERROR;
+    }
+};
+
+class AttestationBase
+{
+public:
+    AttestationBase(ByteSpan nonce) : mNonce(nonce) {}
+    AttestationBase() = delete;
+
+    class CapacityBoundNonce
+    {
+    public:
+        CapacityBoundNonce(ByteSpan nonce) { Set(nonce); }
+        CapacityBoundNonce() = delete;
+
+        const ByteSpan Get() const { return ByteSpan(mNonce, mNonce.Length()); }
+        void Clear() { mNonce.SetLength(0); };
+        void Set(ByteSpan nonce)
+        {
+            mNonce.SetLength(std::min(nonce.size(), mNonce.Capacity()));
+            memmove(mNonce, nonce.data(), mNonce.Length());
+        }
+
+    private:
+        Crypto::CapacityBoundBuffer<kAttestationNonceLength> mNonce;
+    };
+
+    auto Nonce() { return &mNonce; }
+    auto Challenge() { return &mChallenge; }
+    auto Signature() { return &mSignature; }
+    auto Pai() { return &mPai; }
+    auto Dac() { return &mDac; }
+
+private:
+    CapacityBoundNonce mNonce;
+    CapacityBoundStaticBuffer<kAES_CCM128_Key_Length> mChallenge;
+    CapacityBoundSharedBuffer<kMax_ECDSA_Signature_Length> mSignature;
```

Comment:
They are using shared_ptrs so that when the last state or event drops its reference, they are freed.  It's a way to have the state machine automatically manage memory for these dynamically allocated things that only need to exist in certain states.

There is a good, related discussion here: https://github.com/project-chip/connectedhomeip/pull/13294#discussion_r779176030